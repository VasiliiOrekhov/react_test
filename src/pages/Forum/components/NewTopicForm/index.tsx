import { Button, ButtonVariant } from '@components/Button';
import { SubmitHandler, useForm } from 'react-hook-form';
import { Input } from '@components/Input';

import { Spacer } from '@/components';

import s from './index.module.scss';

type NewTopicFormProps = {
	onClick(): void;
};

type FormValues = {
	topicName: string;
};

export const NewTopicForm = ({ onClick }: NewTopicFormProps) => {
	const {
		register,
		handleSubmit,
		formState: { errors }
	} = useForm<FormValues>();

	const onSubmit: SubmitHandler<FormValues> = () => {
		// todo: add handler
		console.log('created');
		onClick();
	};

	// todo: change to form component
	return (
		<form className={s.newTopicForm} onSubmit={handleSubmit(onSubmit)}>
			<Spacer direction="column" gap="40" fullWidth>
				<Input
					error={errors?.topicName}
					{...register('topicName', {
						required: true
					})}>
					Theme name
				</Input>
				<Spacer align="center" direction="column" gap="35">
					<Button type="submit" variant={ButtonVariant.TEXT}>
						Confirm
					</Button>
					<Button variant={ButtonVariant.TEXT} onClick={onClick}>
						Cancel
					</Button>
				</Spacer>
			</Spacer>
		</form>
	);
};
