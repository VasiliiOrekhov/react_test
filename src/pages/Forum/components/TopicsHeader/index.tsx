type TopicsHeaderProps = {
	className: string;
};

export const TopicsHeader = ({ className }: TopicsHeaderProps) => {
	return (
		<div className={className}>
			<span>Themes:</span>
			<span>Comments:</span>
		</div>
	);
};
