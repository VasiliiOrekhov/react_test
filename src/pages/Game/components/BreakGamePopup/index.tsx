import React from 'react';
import { Button, ButtonVariant } from '@components/Button';
import { useNavigate } from 'react-router-dom';

import { Modal, Spacer, Text } from '@/components';

import { routerPaths } from '@/constants/routerPaths';

interface BreakPopupProps {
	onClose: () => void;
	setScore: (score: number) => void;
	destroyedEnemiesCount: (() => number) | undefined;
	isOpen?: boolean;
}

export const BreakGamePopup: React.FC<BreakPopupProps> = ({
	onClose,
	setScore,
	destroyedEnemiesCount,
	isOpen = false
}) => {
	const navigate = useNavigate();

	const handleBackClick = () => {
		onClose();
	};
	const handleConfirmClick = () => {
		destroyedEnemiesCount && setScore(destroyedEnemiesCount());
		navigate(`${routerPaths.main}${routerPaths.gameOver}`);
	};

	return (
		<Modal isOpen={isOpen} onClose={handleBackClick}>
			<Spacer align="center" direction="column" gap="50">
				<Text tag="h1" align="center">
					{'Do you really \n want \n to exit?'}
				</Text>
				<Text tag="h1" align="center">
					Progress will be lost
				</Text>
				<Spacer align="center" direction="column" gap="20">
					<Button variant={ButtonVariant.TEXT} onClick={handleConfirmClick}>
						Confirm
					</Button>
					<Button variant={ButtonVariant.TEXT} onClick={handleBackClick}>
						Cancel
					</Button>
				</Spacer>
			</Spacer>
		</Modal>
	);
};
