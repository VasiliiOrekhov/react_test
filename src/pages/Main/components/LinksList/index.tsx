import { useCallback, useEffect, useState } from 'react';
import classNames from 'classnames';
import { useNavigate } from 'react-router-dom';
import { MouseEvent } from 'react';

import { Link, Text } from '@/components';
import { routerPaths } from '@/constants/routerPaths';
import { logOutUser } from '@/store/reducers/user/userActionCreator';
import { useAppDispatch } from '@/store';

import { navLinks } from './constants/navLinks';
import styles from './index.module.scss';

export const LinksList = () => {
	const dispatch = useAppDispatch();
	const [activeLinkId, setActiveLinkId] = useState<number>(0);
	const navigate = useNavigate();

	const exitHandler = async (e: MouseEvent<HTMLAnchorElement>) => {
		e.stopPropagation();
		e.preventDefault();
		dispatch(logOutUser());
		navigate(routerPaths.login);
	};

	const handleKeyDown = useCallback(
		(event: KeyboardEvent) => {
			event.preventDefault();

			switch (event.key) {
				case 'ArrowUp':
					setActiveLinkId(previousId => {
						if (previousId === null || previousId === 0) return navLinks.length - 1;
						return previousId - 1;
					});
					break;

				case 'ArrowDown':
					setActiveLinkId(previousId => {
						if (previousId === null || previousId === navLinks.length - 1) return 0;
						return previousId + 1;
					});
					break;

				case 'Enter': {
					const to = navLinks.find(link => link.id === activeLinkId)?.path;
					if (to) {
						navigate(to);
					}
					break;
				}

				default:
					return;
			}
		},
		[activeLinkId, navigate]
	);

	useEffect(() => {
		document.addEventListener('keydown', handleKeyDown);

		return () => {
			document.removeEventListener('keydown', handleKeyDown);
		};
	}, [handleKeyDown]);

	return (
		<ul className={styles.list}>
			{navLinks.map(({ id, path, text }) => (
				<li key={id}>
					<Link
						to={path}
						className={classNames({ [styles.isActive]: id === activeLinkId })}
						onMouseEnter={() => setActiveLinkId(id)}
						onClick={text === 'Exit' ? exitHandler : undefined}>
						<Text size="l" variant={id === activeLinkId ? 'selected' : 'normal'}>
							{text}
						</Text>
					</Link>
				</li>
			))}
		</ul>
	);
};
