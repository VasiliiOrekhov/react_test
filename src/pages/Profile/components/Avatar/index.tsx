import React, { useState, useRef, useMemo } from 'react';
import { Button } from '@components/Button';
import { createAvatarPath } from '@pages/Profile/components/utils/createAvatarPath';

import { Input, Modal, Spacer, Text } from '@/components';
import { updateUserAvatar } from '@/store/reducers/user/userActionCreator';
import { useAppSelector, userState } from '@/store/selectors';
import { useAppDispatch } from '@/store';

import styles from './index.module.scss';

export const Avatar: React.FC = () => {
	const dispatch = useAppDispatch();
	const { user, isLoading, error: userError } = useAppSelector(userState);

	const [isPopupOpen, setPopupOpen] = useState(false);
	const [preview, setPreview] = useState<string | null>(null);
	const [imageFile, setImageFile] = useState<File | null>(null);
	const [saveImageError, setSaveImageError] = useState<boolean | null>(null);
	const avatarRef = useRef<HTMLDivElement>(null);
	const fileInputRef = useRef<HTMLInputElement>(null);

	const avatarUrl = () => {
		if (user?.avatar) {
			return createAvatarPath(user?.avatar);
		}

		return '';
	};

	const handleAvatarClick = () => {
		setPopupOpen(true);
	};

	const handleImageClick = () => {
		if (fileInputRef.current) {
			fileInputRef.current.click();
		}

		setSaveImageError(null);
	};

	const handleSaveImage = async () => {
		if (imageFile) {
			try {
				dispatch(updateUserAvatar(imageFile));

				console.log('File uploaded successfully');
			} catch (error) {
				setSaveImageError(true);
				console.error('Error during file upload:', error);
			}
		}
	};

	const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const file = e.target.files?.[0];

		if (file) {
			setImageFile(file);
			const fileUrl = URL.createObjectURL(file);
			setPreview(fileUrl);
		}
	};

	const handleClosePopup = (e?: React.MouseEvent) => {
		e?.stopPropagation();

		setPopupOpen(false);
		setImageFile(null);
		setPreview(null);
		setSaveImageError(null);
	};

	return (
		<div ref={avatarRef} onClick={handleAvatarClick}>
			<div className={styles.avatarMock} style={{ backgroundImage: `url(${avatarUrl()})` }} />
			<Modal isOpen={isPopupOpen} className={styles.modal} onClose={handleClosePopup}>
				<Spacer direction="column" fullWidth>
					<Spacer className={styles.imagePlaceholder} fullWidth>
						{preview ? (
							<img
								className={styles.preview}
								src={preview}
								alt="Изображение: Предпросмотр фотографии пользователя"
							/>
						) : (
							<Text align="center" size="s">
								{'Your image \n will be here'}
							</Text>
						)}
					</Spacer>
					{saveImageError && (
						<Spacer spaceTop="20">
							<Text align="left" size="xs" variant="error">
								An error has occurred. Please try again or come back later
							</Text>
						</Spacer>
					)}
					{userError?.reason && (
						<Text size="s" variant="error">
							{userError?.reason}
						</Text>
					)}
					<Input
						ref={fileInputRef}
						className={styles.avatarInput}
						type="file"
						accept=".jpeg, .jpg, .png, .gif, .webp"
						// @ts-ignore
						onChange={handleFileChange}
					/>
					<Spacer direction="column" fullHeight fullWidth gap="50" spaceTop="40">
						{preview ? (
							<Spacer direction="column" gap="20" fullWidth>
								<Button disabled={isLoading} fullWidth onClick={handleSaveImage}>
									<Text align="center" size="s">
										Confirm
									</Text>
								</Button>
								<Button fullWidth onClick={handleImageClick}>
									<Text align="center" size="s">
										Choose another
									</Text>
								</Button>
							</Spacer>
						) : (
							<Button fullWidth onClick={handleImageClick}>
								<Text align="center" size="s">
									Choose Image
								</Text>
							</Button>
						)}
						<Button fullWidth onClick={handleClosePopup}>
							<Text align="center" size="s">
								Cancel
							</Text>
						</Button>
					</Spacer>
				</Spacer>
			</Modal>
		</div>
	);
};
