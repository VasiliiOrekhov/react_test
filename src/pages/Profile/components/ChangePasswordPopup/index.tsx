import React, { useState, ChangeEvent } from 'react';

import ChangePasswordService from '@/services/changePasswordService';
import { Button, Input, Spacer, Text, FormCard, Modal } from '@/components';

interface ChangePasswordPopupProps {
	onClose: () => void;
	isOpen?: boolean;
}

export const ChangePasswordPopup: React.FC<ChangePasswordPopupProps> = ({
	onClose,
	isOpen = false
}) => {
	const [oldPassword, setOldPassword] = useState('');
	const [newPassword, setNewPassword] = useState('');

	const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
		const { name, value } = e.target;
		if (name === 'oldPassword') {
			setOldPassword(value);
		} else if (name === 'newPassword') {
			setNewPassword(value);
		}
	};

	const handleSaveClick = async () => {
		const changePasswordService = new ChangePasswordService();

		try {
			await changePasswordService.changePassword(oldPassword, newPassword);
			console.log('Password changed successfully');
		} catch (error) {
			console.error('Error changing password:', error);
		}
	};

	const handleBackClick = () => {
		onClose();
	};

	const handlePopupClick = (e: React.MouseEvent) => {
		e.stopPropagation();
	};

	return (
		<Modal isOpen={isOpen} onClose={onClose}>
			<FormCard
				fullWidthContent
				footer={
					<Spacer justify="between" spaceTop="20" gap="80">
						<Button type="button" onClick={handleSaveClick}>
							<Text align="center" size="s">
								Save
							</Text>
						</Button>
						<Button type="button" onClick={handleBackClick}>
							<Text align="center" size="s">
								Back
							</Text>
						</Button>
					</Spacer>
				}>
				<Spacer direction="column" gap="40" fullWidth>
					{/* @ts-ignore */}
					<Input type="password" name="oldPassword" onChange={handleInputChange}>
						old password
					</Input>
					{/* @ts-ignore */}
					<Input type="password" name="newPassword" onChange={handleInputChange}>
						new password
					</Input>
				</Spacer>
			</FormCard>
		</Modal>
	);
};
