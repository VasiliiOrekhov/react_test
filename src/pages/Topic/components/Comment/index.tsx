import { useMemo } from 'react';
import classNames from 'classnames';
import { CommentModel } from '@models/topics';

import { Spacer } from '@/components';

import s from './index.module.scss';

import { CURRENT_USER_ID } from '../../lib/constants';
import { USERS } from '../../lib/mocks';
import { SmileMenu } from '../SmileMenu';

type CommentProps = CommentModel;

export const Comment = (props: CommentProps) => {
	const { userId, messages } = props;
	const commentAuthor = USERS.find(user => user.id === userId);
	const isMainComment = commentAuthor?.id === CURRENT_USER_ID;
	const authorName = useMemo(() => {
		return commentAuthor?.name ?? 'UNKNOWN';
	}, []);

	const mods = {
		[s.currentUserComment]: isMainComment
	};

	return (
		<Spacer direction="column" className={s.commentsWrapper}>
			{messages.map(({ id, text }, i) => (
				<Spacer direction="column" align="start" key={id} className={classNames(s.comment, mods)}>
					{i === 0 && <span className={s.author}>{authorName}</span>}
					<span className={s.text}>{text}</span>
					{!isMainComment && <SmileMenu />}
				</Spacer>
			))}
		</Spacer>
	);
};
