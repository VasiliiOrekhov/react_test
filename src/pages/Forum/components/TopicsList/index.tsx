import React, { useCallback, useEffect } from 'react';
import { Link } from '@components/Link';
import { TopicModel } from '@models/topics';
import { TOPICS_LIST } from '@pages/Forum/lib/mocks';

import { abbreviateNumber } from '@/utils/abbreviateNumber';
import { Text } from '@/components';
import { getTopicsList, getTopic } from '@/store/reducers/forum/forumReducer';
import { forumState, useAppSelector } from '@/store/selectors';
import { useAppDispatch } from '@/store';

import s from './index.module.scss';

import { TopicsHeader } from '../TopicsHeader';

export const TopicsList = () => {
	const { topics, error: forumError, isLoading } = useAppSelector(forumState);
	const dispatch = useAppDispatch();

	useEffect(() => {
		dispatch(getTopicsList(TOPICS_LIST));
	}, []);

	const handleClick = useCallback(
		(topicId: TopicModel['id']) => {
			const selectedTopic = topics?.find(topic => topic.id === topicId);

			if (selectedTopic) {
				dispatch(getTopic(selectedTopic));
			}
		},
		[topics]
	);

	if (isLoading) {
		return (
			<Text align="center" size="m">
				Loading...
			</Text>
		);
	}

	if (forumError?.reason) {
		return (
			<Text size="s" variant="error">
				{forumError?.reason}
			</Text>
		);
	}

	return (
		<div className={s.topics}>
			<TopicsHeader className={s.topic} />
			<ul className={s.topicsList} aria-label="forum-topics-list">
				{topics?.map(topic => (
					<li key={topic.id} className={s.topicItem} title={topic.name}>
						<Link
							to={`topic/${topic.id}`}
							state={{
								topicName: topic.name
							}}
							className={s.topic}
							onClick={() => handleClick(topic.id)}>
							<span className={s.topicName}>{topic.name}</span>
							<span>{abbreviateNumber(topic.length, 0)}</span>
						</Link>
					</li>
				))}
			</ul>
		</div>
	);
};
