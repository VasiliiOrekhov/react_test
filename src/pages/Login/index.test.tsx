import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import { TestWrapper } from '@/test/components/TestWrapper';

import { Login } from './';

describe('Login Page', () => {
	let usernameInput: HTMLInputElement;
	let passwordInput: HTMLInputElement;

	beforeEach(() => {
		render(
			<TestWrapper>
				<Login />
			</TestWrapper>
		);

		usernameInput = screen.getByTestId('login_input_first-name');
		passwordInput = screen.getByTestId('login_input_password');
	});

	test('should render error text for login & password on blur', async () => {
		// Пользователь вводит значение из двух символов. Минимум - 3
		fireEvent.change(usernameInput, { target: { value: 'ab' } });
		fireEvent.blur(usernameInput);

		// Пользователь вводит значение из шести символов. Минимум - 8
		fireEvent.change(passwordInput, { target: { value: '312dad' } });
		fireEvent.blur(passwordInput);

		await waitFor(() => {
			expect(screen.getByText('Длина логина должна быть от 3 до 20 символов')).toBeDefined();
			expect(screen.getByText('Длина пароля должна быть от 8 до 40 символов')).toBeDefined();
		});
	});

	test('should render error text for login & password on submit', async () => {
		const submitButton = screen.getByText('Sign In');

		// Поле не должно содержать спец символы
		fireEvent.change(usernameInput, { target: { value: 'dsad@!' } });

		// Пользователь вводит значение без чисел. Необходимо хотя бы одно
		fireEvent.change(passwordInput, { target: { value: 'abcdefgh' } });

		fireEvent.click(submitButton);

		await waitFor(() => {
			expect(screen.getByText('Поле не должно содержать спецсимволы')).toBeDefined();
			expect(screen.getByText('В пароле должно быть хотя бы одно число')).toBeDefined();
		});
	});

	test('should render default', () => {
		const { container } = render(
			<TestWrapper>
				<Login />
			</TestWrapper>
		);

		expect(container).toMatchSnapshot();
	});
});
