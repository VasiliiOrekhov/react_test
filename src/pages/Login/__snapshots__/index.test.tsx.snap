// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render default 1`] = `
<div>
  <main>
    <div
      class=""
    >
      <div
        class=""
      >
        <p
          class=""
        >
          Galaxy 
 Conquerors
        </p>
        <section
          class=""
        >
          <p
            class=""
          >
            Authorization
          </p>
          <div
            class=""
          >
            <form>
              <div
                class=""
              >
                <label
                  for="login"
                >
                  login
                </label>
                <input
                  class=""
                  data-testid="login_input_first-name"
                  name="login"
                  style="height: 0px;"
                  type="text"
                  value=""
                />
              </div>
              <div
                class=""
              >
                <label
                  for="password"
                >
                  password
                </label>
                <input
                  class=""
                  data-testid="login_input_password"
                  name="password"
                  style="height: 0px;"
                  type="password"
                  value=""
                />
              </div>
            </form>
          </div>
          <div
            class=""
          >
            <div
              class=""
            >
              <button
                class=""
                type="submit"
              >
                <span />
                Sign In
              </button>
            </div>
          </div>
        </section>
      </div>
      <a
        class=""
        href="/registration"
      >
        <div
          class=""
        >
          <p
            class=""
          >
            x
          </p>
          <p
            class=""
          >
            Don not have an account yet?
          </p>
        </div>
      </a>
    </div>
  </main>
</div>
`;
